src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/nestedfield/NestedFieldProjection.java

```diff
@@ -123,23 +135,64 @@ private void buildInnerHit(List<String> fieldNames, NestedQueryBuilder query) {
+
         return nestedQueries.stream().
-                             filter(query -> isSamePath(path, query)).
-                             findAny().
-                             orElseGet(createEmptyNestedQuery(path));
+                filter(query -> isSamePath(path, query, fieldNames)).
+                findAny().
+                orElseGet(createNestedQuery(path, fieldNames));
+    }
+
+    /** Check against NestedQuery of WHERE condition */
+    private boolean isSamePath(String path, NestedQueryBuilder query, List<String> fieldNames) {
+        boolean isSamePath = nestedQuery(path, query.query(), query.scoreMode()).equals(query);
+        if (isSamePath) {
+           setLastValues(path, fieldNames);
+        }
+        return isSamePath;
     }
 
-    private boolean isSamePath(String path, NestedQueryBuilder query) {
-        return nestedQuery(path, query.query(), query.scoreMode()).equals(query);
+    /**
+     * Since groupFieldNamesByPath() uses TreeMap to maintain reverse alphabetical order. Here we maintain previous
+     * path and fieldName value to compare with current nested query.
+     * */
+    private boolean hasSameParent(String path, QueryBuilder query) {
+        // query could also be termQueryBuilder
+        if (!lastFieldInPath.isEmpty() && (query instanceof NestedQueryBuilder)) {
+            NestedQueryBuilder nestedQuery = (NestedQueryBuilder) query;
+
+            String finalPath = path + "." + lastFieldInPath;
+            NestedQueryBuilder tmp = nestedQuery(finalPath, nestedQuery.query(), nestedQuery.scoreMode());
+            buildInnerHit(lastFieldNames, tmp);
+            return tmp.equals(query);
+        }
+        return false;
     }
 
     /** Create a nested query with match all filter to place inner hits */
-    private Supplier<NestedQueryBuilder> createEmptyNestedQuery(String path) {
+    private Supplier<NestedQueryBuilder> createNestedQuery(String path, List<String> fieldNames) {
         return () -> {
-            NestedQueryBuilder nestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
-            ((BoolQueryBuilder) query().filter().get(0)).must(nestedQuery);
-            return nestedQuery;
+            NestedQueryBuilder newNestedQuery;
+            List<QueryBuilder> mustList = ((BoolQueryBuilder) query().filter().get(0)).must();
+            int lastIndex = mustList.size() - 1;
+
+            if (mustList.isEmpty() || !hasSameParent(path, mustList.get(lastIndex))) {
+                // create empty nested query
+                newNestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
+            } else {
+                // create a nested query containing another nested query
+                NestedQueryBuilder oldNestedQuery = (NestedQueryBuilder) mustList.get(lastIndex);
+                newNestedQuery = nestedQuery(path, oldNestedQuery, ScoreMode.None);
+
+                // remove outdated nested query
```

Comment:
np: is it outdated or already processed?

Reply:
The logic here is created a new nested query containing this existing nested query. Then we remove that existing nested query from `request`. Finally we add a new nested query block to the must domain ` ((BoolQueryBuilder) query().filter().get(0)).must(newNestedQuery);` So It is outdated because we have generated a new one to replace his position.